src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ChannelHolder.java

```diff
@@ -0,0 +1,246 @@
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.rabbitmq.client.AMQP.BasicProperties;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.ConfirmListener;
+import com.rabbitmq.client.ShutdownNotifier;
+
+class ChannelHolder implements ConfirmListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChannelHolder.class);
+    private static final Consumer<Map<Long, PublicationHolder>> DO_NOTHING = it -> {};
+    private final Lock lock = new ReentrantLock();
+    private final Supplier<Channel> supplier;
+    private final BiConsumer<ShutdownNotifier, Boolean> reconnectionChecker;
+    private final int maxCount;
+    private final boolean retryOnPublishNotConfirmed;
+
+    private final NavigableMap<Long, PublicationHolder> publishedData;
+
+    @GuardedBy("lock")
+    private int pending;
+    @GuardedBy("lock")
+    private Future<?> check;
+    @GuardedBy("lock")
+    private Channel channel;
+
+    public ChannelHolder(
+            Supplier<Channel> supplier,
+            BiConsumer<ShutdownNotifier, Boolean> reconnectionChecker,
+            int maxCount,
+            boolean retryOnPublishNotConfirmed
+    ) {
+        this.supplier = Objects.requireNonNull(supplier, "'Supplier' parameter");
+        this.reconnectionChecker = Objects.requireNonNull(reconnectionChecker, "'Reconnection checker' parameter");
+        this.maxCount = maxCount;
+        this.retryOnPublishNotConfirmed = retryOnPublishNotConfirmed;
+        publishedData = retryOnPublishNotConfirmed ? new ConcurrentSkipListMap<>() : Collections.emptyNavigableMap();
+    }
+
+    public void withLock(ChannelConsumer consumer) throws IOException {
+        withLock(true, consumer);
+    }
+
+    public void publish(String exchange, String routingKey, BasicProperties props, byte[] body) throws IOException {
+        withLock(channel -> {
+            if (retryOnPublishNotConfirmed) {
+                publishedData.put(channel.getNextPublishSeqNo(),
+                        new PublicationHolder(exchange, routingKey, props, body));
+            }
+            channel.basicPublish(exchange, routingKey, props, body);
+        });
+    }
+
+    private void publish(PublicationHolder holder) throws IOException {
+        publish(holder.getExchange(), holder.getRoutingKey(), holder.getProps(), holder.getBody());
+    }
+
+    public void withLock(Runnable action) {
```

Comment:
Need for this method seems questionable to me ATM because all its usages call methods an holder which are already locked. If we still want such method then maybe replace with the one which accept holder consumer instead?